File: server/src/services/scoring.ts
import { createHash } from "crypto";

const AI_WEIGHT_LOW = Number(process.env.AI_WEIGHT_LOW_VOTES ?? 0.95);
const AI_WEIGHT_HIGH = Number(process.env.AI_WEIGHT_HIGH_VOTES ?? 0.75);
const VOTE_LOW = Number(process.env.VOTE_THRESHOLD_LOW ?? 10);
const VOTE_HIGH = Number(process.env.VOTE_THRESHOLD_HIGH ?? 100);
const MAX_URL_LENGTH = Number(process.env.MAX_URL_LENGTH ?? 2048);

export function normalizeUrl(url: string): string {
  const u = new URL(url);
  u.hash = "";
  let path = u.pathname.replace(/\/+$/, "") || "/";
  u.pathname = path;
  return u.toString().toLowerCase();
}

export function hashUrl(url: string): string {
  return createHash("sha256").update(normalizeUrl(url)).digest("hex");
}

export function extractDomain(url: string): string {
  return new URL(url).hostname.toLowerCase();
}

export function validateUrl(url: string): void {
  if (!url || !url.trim()) throw new Error("URL must not be empty.");
  if (url.length > MAX_URL_LENGTH)
    throw new Error(`URL exceeds the maximum length of ${MAX_URL_LENGTH} characters.`);
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new Error("Invalid URL format.");
  }
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:")
    throw new Error("URL must use the http or https scheme.");
  if (!parsed.hostname) throw new Error("URL must include a host.");
}

const VOTE_SCORES: Record<string, number> = { human: 0, mixed: 0.5, ai_generated: 1 };

export function calculateCrowdScore(
  votes: Array<{ vote: string; reputation: number }>
): number | null {
  if (votes.length === 0) return null;
  let wSum = 0, wTotal = 0;
  for (const { vote, reputation } of votes) {
    const w = Math.max(reputation, 0.1);
    wSum += (VOTE_SCORES[vote] ?? 0.5) * w;
    wTotal += w;
  }
  return wTotal === 0 ? null : wSum / wTotal;
}

export function calculateCombinedScore(
  aiScore: number | null,
  crowdScore: number | null,
  voteCount: number
): number | null {
  if (aiScore === null && crowdScore === null) return null;
  if (aiScore === null) return crowdScore;
  if (crowdScore === null) return aiScore;

  let aiWeight: number;
  if (voteCount <= VOTE_LOW) aiWeight = AI_WEIGHT_LOW;
  else if (voteCount >= VOTE_HIGH) aiWeight = AI_WEIGHT_HIGH;
  else {
    const ratio = (voteCount - VOTE_LOW) / (VOTE_HIGH - VOTE_LOW);
    aiWeight = AI_WEIGHT_LOW - ratio * (AI_WEIGHT_LOW - AI_WEIGHT_HIGH);
  }
  return aiWeight * aiScore + (1 - aiWeight) * crowdScore;
}

export function scoreToConfidence(voteCount: number, aiScore: number | null): string {
  if (voteCount === 0 && aiScore === null) return "none";
  if (voteCount < 5) return "low";
  if (voteCount < 50) return "medium";
  return "high";
}

export function scoreToVerdict(score: number | null): string {
  if (score === null) return "unknown";
  if (score <= 0.3) return "human";
  if (score <= 0.7) return "mixed";
  return "ai_generated";
}

export function detectPlatform(url: string): string {
  let host: string;
  try {
    host = new URL(url).hostname.toLowerCase().replace(/^www\./, "");
  } catch {
    return "generic";
  }
  if (host === "instagram.com" || host.endsWith(".instagram.com")) return "instagram";
  if (host === "facebook.com" || host === "fb.com" || host === "fb.watch" || host.endsWith(".facebook.com")) return "facebook";
  if (host === "reddit.com" || host === "redd.it" || host.endsWith(".reddit.com")) return "reddit";
  if (host === "youtube.com" || host === "youtu.be" || host.endsWith(".youtube.com")) return "youtube";
  if (host === "tiktok.com" || host.endsWith(".tiktok.com")) return "tiktok";
  if (host === "x.com" || host === "twitter.com" || host.endsWith(".x.com") || host.endsWith(".twitter.com")) return "twitter";
  return "generic";
}

File: backend/app/services/scoring.py
from __future__ import annotations

import hashlib
from urllib.parse import urlparse, urlunparse

from app.config import settings


def validate_url(url: str) -> None:
    """Validate a user-supplied URL, raising ValueError if unacceptable."""
    if not url or not url.strip():
        raise ValueError("URL must not be empty.")
    if len(url) > settings.MAX_URL_LENGTH:
        raise ValueError(
            f"URL exceeds the maximum length of {settings.MAX_URL_LENGTH} characters."
        )
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https"):
        raise ValueError("URL must use the http or https scheme.")
    if not parsed.netloc:
        raise ValueError("URL must include a host.")


def normalize_url(url: str) -> str:
    parsed = urlparse(url)
    normalized = urlunparse(
        (
            parsed.scheme.lower(),
            parsed.netloc.lower(),
            parsed.path.rstrip("/") or "/",
            parsed.params,
            parsed.query,
            "",
        )
    )
    return normalized


def hash_url(url: str) -> str:
    normalized = normalize_url(url)
    return hashlib.sha256(normalized.encode()).hexdigest()


def extract_domain(url: str) -> str:
    return urlparse(url).netloc.lower()


VOTE_SCORES = {"human": 0.0, "mixed": 0.5, "ai_generated": 1.0}


def calculate_crowd_score(
    votes: list[tuple[str, float]],
) -> float | None:
    """Calculate crowd score from list of (vote_type, voter_reputation) tuples."""
    if not votes:
        return None

    weighted_sum = 0.0
    weight_total = 0.0
    for vote_type, reputation in votes:
        weight = max(reputation, 0.1)
        weighted_sum += VOTE_SCORES.get(vote_type, 0.5) * weight
        weight_total += weight

    if weight_total == 0:
        return None
    return weighted_sum / weight_total


def calculate_combined_score(
    ai_score: float | None,
    crowd_score: float | None,
    vote_count: int,
) -> float | None:
    if ai_score is None and crowd_score is None:
        return None
    if ai_score is None:
        return crowd_score
    if crowd_score is None:
        return ai_score

    low = settings.VOTE_THRESHOLD_LOW
    high = settings.VOTE_THRESHOLD_HIGH
    w_ai_low = settings.AI_WEIGHT_LOW_VOTES
    w_ai_high = settings.AI_WEIGHT_HIGH_VOTES

    if vote_count <= low:
        ai_weight = w_ai_low
    elif vote_count >= high:
        ai_weight = w_ai_high
    else:
        ratio = (vote_count - low) / (high - low)
        ai_weight = w_ai_low - ratio * (w_ai_low - w_ai_high)

    crowd_weight = 1.0 - ai_weight
    return ai_weight * ai_score + crowd_weight * crowd_score


def score_to_confidence(vote_count: int, ai_score: float | None) -> str:
    if vote_count == 0 and ai_score is None:
        return "none"
    if vote_count < 5:
        return "low"
    if vote_count < 50:
        return "medium"
    return "high"
